implementation of nameserver

starius opened the pull request:
Current implementation is very basic DNS server responding with random IP from pool. It is based on tutorial http://mkaczanowski.com/golang-build-dynamic-dns-service-go/

I think, the nameserver should be replicated not to be a weak spot of the system. End users specify multiple nameservers for a domain. [etcd](https://github.com/coreos/etcd) can be used to store and synchronize information (list of entry proxies with additional data).

I think, entry proxies should register automatically in nameserver(s) using `nsupdate`.

starius committed:
add draft of nameserver

Based on tutorial http://mkaczanowski.com/golang-build-dynamic-dns-service-go/

starius committed:
nameserver: split resolver from request handler

And add argument Qclass to the main function of resolver.

starius committed:
nameserver: handle TXT requests

see https://github.com/DonnchaC/oniongateway/issues/18

starius committed:
nameserver: factor out FixedResolver, add test

starius committed:
nameserver: add etcd_resolver

starius committed:
nameserver uses fixed- and etcd resolver

starius committed:
nameserver: log whole DNS message

... instead of only query section

starius committed:
nameserver: fix golint warnings

starius committed:
etcd_resolver_test: sleep before creating client

Tests failed with the following error:

2016-08-21 16:25:30.299559 I | etcdserver: published {Name:default ClientURLs:[http://127.0.0.1:38202]} to cluster f12d0b8c5b4beff9
2016-08-21 16:25:30.299680 I | embed: ready to serve client requests
2016-08-21 16:25:30.299769 I | etcdserver: setting up the initial cluster version to 3.0
2016-08-21 16:25:30.301468 N | membership: set the initial cluster version to 3.0
2016-08-21 16:25:30.303932 N | embed: serving insecure client requests on 127.0.0.1:38202, this is strongly discouraged!
2016-08-21 16:25:30.320528 I | api: enabled capabilities for version 3.0
2016-08-21 16:25:30.320623 W | etcdserver: apply entries took too long [19.219094ms for 1 entries]
2016-08-21 16:25:30.320676 W | etcdserver: avoid queries with large range/delete range!
2016-08-21 16:25:30.320718 I | etcdserver: skipped leadership transfer for single member cluster
--- FAIL: TestEtcdResolver (0.73s)
        etcd_resolver_test.go:104: Failed to populate etcd with example data: Failed to put key /ipv4/127.0.0.1: etcdserver: not capable

Discussion [1] suggests that there is a race in etcd during first 500 ms.
Sleep of one second should fix the problem.

[1] https://github.com/coreos/etcd/issues/6029#issuecomment-234664336

starius commented:
Updated.

Current implementation (f58d5b9e0c3a85273bd05e7ed99a4aa89a91a52a) provides static config file or etcd as a backend.

How to use:

**Static config file**:

```
nameserver -static-config static_config.yaml -listen-addr :4253
```

**Etcd**:

Build etcd and etcdctl, start etcd:

```
go get github.com/coreos/etcd
go get github.com/coreos/etcd/etcdctl
etcd
```

Populate etcd with sample records:

```
ETCDCTL_API=3 etcdctl put /ipv4/94.102.53.178 dummy
ETCDCTL_API=3 etcdctl put /ipv6/2002:5e66:35b2::1 dummy
ETCDCTL_API=3 etcdctl put /domain2onion/pasta.cf. pastagdsp33j7aoq.onion
```

Run `nameserver`:

```
nameserver -etcd-endpoints 127.0.0.1:2379 -listen-addr :4253
```

**Test**:

```
dig @127.0.0.1 -p 4253 pasta.cf A
dig @127.0.0.1 -p 4253 pasta.cf AAAA
dig @127.0.0.1 -p 4253 pasta.cf TXT
```

starius committed:
travis: disable Go 1.3 and Go 1.4

github.com/grpc/grpc-go requires Go >= 1.5

starius committed:
nameserver: factor out handler.go and add test

starius committed:
nameserver: Resolve returns multiple results

starius committed:
nameserver: implement returning multiple results

starius committed:
nameserver: s/fixed resolver/static resolver/

DonnchaC commented:
I have been working on a web interface and API to allow users to easily sign-up for a DNS resolver server. Currently the API is working and I can generate BIND-style zone files with domain and proxy data that is stored in the database from the API.

My plan is to signup for a public authoritative DNS provider and use their infrastructure to provide redundant and geographically distributed DNS hosting.  As the service will generate BIND-style zone files it should be easy for other people to use it with their own DNS setup. 

https://github.com/DonnchaC/oniongate-web

starius committed:
etcd_resolver_test: change config while test

starius commented:
> My plan is to signup for a public authoritative DNS provider and use their infrastructure to provide redundant and geographically distributed DNS hosting.

IMHO, this approach is not ideal:
- dependency on the DNS provider used. It would be a single point of failure: if the DNS provider decides to suspend our account, the service will stop operating. I would prefer deploying a bunch of DNS servers in different VPS providers to distribute this risk.
- it requires a user to sign up for our service (one more step to do!)
- full list of used domains and onion sites is collected in the zone file. I think this should be avoided for the same reasons the Tor project prevents gathering full list of onion sites.

The real problem with current approach is that the user can't set `IN NS` and `TXT` at the same time. (I am not sure - I do not know DNS internals so deep.) I have a crazy idea: can onion site be set in a subdomain of NS server itself:

```
pasta.cf IN NS pastagdsp33j7aoq.onion.our-cool-authoritative-dns-server.com
```

?

starius commented:
TODO: return `status: SERVFAIL` if a domain was not found in `/domain2onion/`

DonnchaC commented:
I don't understand how the trust-relationship would work with a a distributed set of DNS servers. How do onion service operators authenticate updates to the domain -> service mapping? 

When the DNS records are distributed in etcd is there one authoritative node which can make changes or can all DNS servers update the records served by the others?

starius committed:
docs: add diagram of requests routing

routing.svf was generated from routing.dot:

    dot routing.dot -Tsvg -o routing.svg

routing.svg was edited manually in Inkscape.
Result: https://gist.github.com/409007930cc8e1e11edd401ae14c9435

Then it was converted to routing.png

starius commented:
I have added a diagram of requests flow.

![routing](https://i.imgur.com/u1O0KUv.png)

> How do onion service operators authenticate updates to the domain -> service mapping?

Onion service operators change NS records of a domain in registrar.

Before: `pasta.cf. IN NS pastagdsp33j7aoq.onion.oniongw.com.`
After: `pasta.cf. IN NS hss3uro2hsxfogfq.onion.oniongw.com.`

> When the DNS records are distributed in etcd is there one authoritative node which can make changes or can all DNS servers update the records served by the others?

It is to be decided. `nameserver` needs only read-only access to etcd, so a system which changes etcd can be separated from our nameservers.

starius committed:
make etcd_resolver_test more verbose

starius committed:
nameserver: use watcher to get updates from etcd

starius committed:
etcd_resolver: watch from current_rev+1 revision

If etcd is watched from current_rev, last event is applied twice.

starius committed:
etcd_resolver: add logging for etcd watcher

starius committed:
etcd_resolver: remove code fetching initial state

Instead watch from revision=1, which reads all history
from the beginning.

starius committed:
etcd_resolver: fix O(N) removal of address

Keep map[string]index to get index of removed address in O(1).

starius committed:
etcd_resolver_test: print error message on failure

starius committed:
nameserver: do not use time.Sleep in handler_test

starius committed:
etcd_resolver_test: replace sleep with channels

starius committed:
etcd_resolver_test: timeout is 10s

starius commented:
`/domain2onion/*` section of etcd and TXT responses of the DNS server are needed to support https://github.com/DonnchaC/oniongateway/issues/18
